### ISSUE TITLE
Remove CrossWire extension from ASP.NET Core integration package

### ISSUE THREAD

**Issue by dotnetjunkie** [simpleinjector/SimpleInjector#261]
The current implementation of the `CrossWire` extension method of the ASP.NET Core integration package is flawed. Internally it resolves the cross-wired service from the 'application services'. When resolving an instance through the 'application services' you'll get a singleton; even if that component is registered as scoped. No exception is thrown and the application fails silently. The results are quite similar to what happens when having [Captive Dependencies](http://blog.ploeh.dk/2014/06/02/captive-dependency/).

So you should always resolve from the 'request services' (which is the ASP.NET equivalent of resolving from a scope). To work correctly, CrossWire will have to resolve from the 'request services'. Problem with this is that 'request services' are resolved from the HttpContext, but this will not be available from a background thread. According to [this answer](https://stackoverflow.com/questions/27122427/) the `HttpContext.RequestServices` property might even return `null` when some configuration is missing. So getting the CrossWire right can actually be quite tricky.

Because of the problems in the current implementation of CrossWire, and since CrossWire does not promote a SOLID way of application designing, I opt for removing the CrossWire extension method from the library. This prevents our users from getting into trouble by this method and pushes them in to a practice that involves the use of application specific abstractions and adapter implementations that adapt application code to framework and 3rd party components.

**Comment by Tornhoof** [comment]
I think you're right to remove it, but on the other hand I'm not certain that this will work.
For all the internal ASP.NET core services which the user will never see, you do not need to cross wire it.
For all the normal business style services you usually  don't need the internal core services.
But the most obvious counter example is ILoggerFactory, you probably want to have the same logging mechanism for both the internal api and the business code, so you need to "crosswire" ILoggerFactory (either manually with AddSingleton and use a resolved instance from simpleinject or using your crosswire method).

ILoggerFactory is the simple case, no further dependencies to other internal core services in the constructor.
The method unfortunately gets very fast very complex if you want to use other internal services, e.g. API Explorer for documentation (e.g. swashbuckle etc.).

Considering the amount of work necessary to part a typical asp.net project to core, I think it unwise for simpleinject to complicate the porting process even more, as it is rather tedious work to find all the dependencies for the more complex scenarios as docoumentation is currently lacking.

I would probably move it into a compatibility project, as it will be interesting for porting projects from asp.net 4.6 to core.

**Comment by Asshiah** [comment]
Another example of such an internal core services that is likely to be used in the controllers is the `IHtmlStringLocalizerFactory` or the `IStringLocalizerFactory` in order to be able to use the native localization API of the framework.

**Comment by dotnetjunkie** [comment]
I don't agree with your `ILoggerFactory` example. Even if you want to use the built-in logging, you probably still want to hide it behind [your own application specific abstraction](https://stackoverflow.com/questions/5646820/logger-wrapper-best-practice). The adapter for this application-specific abstraction can use ILoggerFactory in the background.

And even if you want to use ILoggerFactory directly, you can do so by cross-wiring it like this:

``` c#
container.RegisterSingleton<ILoggerFactory>(app.ApplicationServices.GetService<ILoggerFactory>());
```
You don't need the `CrossWire` functionality for this at all.

The same holds for `IHtmlStringLocalizerFactory` and `IStringLocalizerFactory`. I would argue hiding them behind your own abstractions, but you can still cross-wire them yourself.

But things obviously get more complicated when dealing with scoped dependencies; which is one of the many reasons that I think that ASP.NET components should always be singleton (which I have discussed before with MS). Problem of course is that the lifestyle by which Microsoft registers its components seems completely arbitrary, but this makes it really easy for developers to accidentally get [Captive Dependencies](http://blog.ploeh.dk/2014/06/02/captive-dependency/) while the built-in container doesn't help at all here.

> I also inject IOptions<SettingsClass> in order to be able to acces application settings.

This is something I would advice you to stop doing. `IOptions<T>` is a badly thought out concept as I expessed [here](https://github.com/simpleinjector/SimpleInjector/issues/143#issuecomment-155029876). You are much better of loading the configuration at app startup and pass it on to the service, as explained in linked article.

Problem with removing CrossWire is of course that developers will have to find out how to do things 'properly'. We should provide them with clear guidance and make sure we don't make their life extraordinary hard.

**Comment by Tornhoof** [comment]
Thank you for your advice, I saw your post for the logging on SO (linked in some older issue).
I'm thinking about a decorator pattern which allows me to resolve the types I want from the asp.net container, e.g. IResolveFromCore<ILoggerFactory> which has a property T Value and if that decorator class is resolved with a higher? lifestyle as the requesting one, i should atleast get validation errors if a singleton registration requests a scope asp.net type. I saw in your documentation some fairly advanced context-based decorators, which might help there for even more complex checking.
